### DISCUSSION TITLE
Placeholder/swallow container, remain on close?

### DISCUSSION THREAD

**Discussion by matthiash82** [i3/i3#5634] [Ideas]
I prefer having a fixed/static layout to divide my large workspace into more manageable units. A typical setup is having one larger container in the middle, with two smaller ones on the side.

It's pretty easy to script i3 to create this layout and get place holder containers where I want applications to open. However, as soon as I close the last application in a container (that I have as tabbed), the container is closed as well and my layout is "destroyed". Using custom scripts for killing and moving windows (moving last window out from a container also closes it) that dynamically opened new placeholders when necessary I could kind of simulate it, but it was never perfect. For example, using `exit` to close a terminal would never call my custom scripts, and the container would be closed, the layout "destroyed". 

I looked into doing something with i3's subscribe feature, but as far as I have been able to tell, the event will be fired too late for my needs.

One idea I had was if i3 could have a feature where a swallow/placeholder-window was _never_ closed. Today, when such a container swallows an application, the container will be closed with the application. Instead I'd want the placeholder to remain / be relaunched. 

An additional feature request would be for such placeholder / swallow windows to _ignore_ floating windows. Is it possible? An easy example of when this gets in the way is if I have a placeholder present (= no swallow yet) and a file dialog window is opened.

**Comment by noncog** [comment]
I would really like to see this feature or some more appropriate means of doing actions like this added to i3. This is my only real complaint with i3 and I've done a lot hacky stuff to simulate workflows like this. For example trying to get a "pop-up" tiling side-bar (potentially with sub-windows and programmed layouts too).

BudLabs on YouTube has gone to great hacky lengths using Bash, Awk, and Python to achieve different variations on this idea and it's really just a hacky mess with limited usability. He's released a lot of tools for this on GitHub as well.

A similar or alternate request that I have is "pinning" or "marking" a container as not to be moved, resized, or closed, perhaps unless by certain "things" with "authority". Extending or relying on the current marking system is also not viable as marks seem to be lost on reload I believe. BudLabs had to develop a wild hack to use marks to encode strings with positional information and several watcher scripts to ensure things are where they're supposed to be and another to encode the marked windows and layout into his own layout file to be re-applied upon reload. I've seen how fragile it is and the current IPC system really can't support this level of control using the means we have.

I really hope we can start some conversation to add some better solutions to the IPC system to support better control like this. I know Stapelberg's views on adding idiosyncratic features to the window manager so I think making a more powerful IPC system is the only viable solution I see here at the moment.

Perhaps some small tweak exists that could afford this without great effort like adding some flags to containers and saving/reloading of state for select containers on reload could do it. Then we could rely on this state information and it's possible extensions like flagging a container to only be a certain layout, position, etc. and not have to re-modify layouts for every automatic change inducing screen flicker with fragile scripts and broken states. Who knows, would love to see it.
